extensions/amp-analytics/0.1/amp-analytics.js

```diff
@@ -0,0 +1,190 @@
+    config = JSON.parse(config);
+    this.config_ = config;
+
+    const type = this.element.getAttribute('type');
+    if (type && AnalyticsConfig[type]) {
+      this.type_ = type;
+
+      // Handle Optout.
+      if (AnalyticsConfig[type]['optout'] &&
+          AnalyticsConfig[type]['optout']()) {
+        // Nothing to do when the user has opted out.
+        return;
+      }
+    }
+
+    this.generateRequests_();
+
+    if (!Array.isArray(config['triggers'])) {
+      // TODO: Nothing to do here. Show error in dev mode.
+      console./*OK*/error(this.getName() + ' No triggers were found in the ' +
+          'config. No analytics data will be sent out.');
+      return;
+    }
+
+    // Trigger callback can be synchronous. Do the registration at the end.
+    for (const trigger of config['triggers']) {
+      if (!trigger['on'] || !trigger['request']) {
+        // TODO: Show warning in dev mode.
+        // TODO: Check for scheme of request.
+        console./*OK*/error(this.getName() + ' "on" and "request" attributes ' +
+            'are required for data to be collected.');
+        continue;
+      }
+      addListener(trigger['on'], this.handleEvent_.bind(this, trigger));
+    }
+  }
+
+  /** @override */
+  buildCallback() {
+    // Remove user defined size. Pixels should always be the default size.
+    this.element.style.width = '';
+    this.element.style.height = '';
+    // Consider the element invisible.
+    this.element.setAttribute('aria-hidden', 'true');
+  }
+
+  /**
+   * Goes through all the requests in predefined vendor config and tag's config
+   * and creates a map of request name to request template. These requests can
+   * then be used while sending a request to a server.
+   */
+  generateRequests_() {
+    const requests = {};
+    for (const config of [AnalyticsConfig[this.type_], this.config_]) {
+      if (!config || !config['requests']) {
+        continue;
+      }
+      for (const name in config['requests']) {
+        if (config['requests'].hasOwnProperty(name)) {
+          requests[name] = config['requests'][name];
+        }
+      }
+    }
+    this.requests_ = requests;
+
+    // Now replace any request templates with their values.
+    const all = Object.keys(this.requests_).join('|');
+    const requestExpr = new RegExp('\{(' + all + ')\}', 'g');
+    for (const k in this.requests_) {
+      this.requests_[k] = this.expandRequest_(this.requests_[k], requestExpr,
+          this.requests_, 5);
+    }
+  }
+
+  /**
+   * Replaces template variables corresponding to a request with the value of
+   * that request. If no request is found, an empty string is inserted instead.
+   */
+  expandRequest_(formatString, expression, requests, depth) {
+    if (depth <= 0) {
+      return '';
+    }
+    const that = this;
+    depth--;
+    const request = formatString.replace(expression, function(match, name) {
+      return that.expandRequest_(requests[name], expression, requests, depth) ||
```

Comment:
For a setup like below:


'requests' : {
  'base': '/something',
  'pageview': '{base}?domain={domain}',
  'pageview_with_custom_dimensiom': '{pageview}&cd1={custom_dimension}'
}

The first two requests could be in predefined type with the third defined in inline config.